### ISSUE TITLE
CrashloopBackOff for capa-controller-manager on v0.6.8

### ISSUE THREAD

**Issue by thof** [kubernetes-sigs/cluster-api-provider-aws#2691]
/kind bug

**What steps did you take and what happened:**
A CrashloopBackOff caused by SIGSEGV when capa-controller-manager fails to reconcile the nodegroup (mettwo-0) for the management cluster (EKS):

```
I0820 12:16:50.190041       1 controller.go:189] controller-runtime/controller "msg"="Starting workers" "controller"="awsmachinepool" "worker count"=1
I0820 12:16:50.190253       1 controller.go:162] controller-runtime/controller "msg"="Starting Controller" "controller"="awsfargateprofile" 
I0820 12:16:50.190289       1 controller.go:189] controller-runtime/controller "msg"="Starting workers" "controller"="awsfargateprofile" "worker count"=5
I0820 12:16:50.190356       1 generic_predicates.go:168] controllers/AWSFargateProfile "msg"="Resource is not paused, will attempt to map resource" "awsmanagedcontrolplane"="mettwo" "namespace"="default" "predicate"="createEvent" 
I0820 12:16:50.281004       1 awsmanagedmachinepool_controller.go:181] controllers/AWSManagedMachinePool "msg"="Reconciling AWSManagedMachinePool" "AWSManagedControlPlane"="mettwo" "AWSManagedMachinePool"="mettwo-0" "Cluster"="mettwo" "MachinePool"="mettwo-0" "namespace"="default" 
I0820 12:16:50.281549       1 eks.go:86] controllers/AWSManagedMachinePool "msg"="Reconciling EKS nodegroup" "AWSManagedControlPlane"="mettwo" "AWSManagedMachinePool"="mettwo-0" "Cluster"="mettwo" "MachinePool"="mettwo-0" "namespace"="default" 
I0820 12:16:50.281583       1 roles.go:155] controllers/AWSManagedMachinePool "msg"="Reconciling EKS Nodegroup IAM Role" "AWSManagedControlPlane"="mettwo" "AWSManagedMachinePool"="mettwo-0" "Cluster"="mettwo" "MachinePool"="mettwo-0" "namespace"="default" 
I0820 12:16:50.434750       1 iam.go:210] controllers/AWSManagedMachinePool "msg"="Ensuring tags and AssumeRolePolicyDocument are set on role" "AWSManagedControlPlane"="mettwo" "AWSManagedMachinePool"="mettwo-0" "Cluster"="mettwo" "MachinePool"="mettwo-0" "namespace"="default" 
I0820 12:16:50.435036       1 iam.go:123] controllers/AWSManagedMachinePool "msg"="Ensuring Polices are attached to role" "AWSManagedControlPlane"="mettwo" "AWSManagedMachinePool"="mettwo-0" "Cluster"="mettwo" "MachinePool"="mettwo-0" "namespace"="default" 
I0820 12:16:50.479280       1 nodegroup.go:43] controllers/AWSManagedMachinePool "msg"="describing eks node group" "AWSManagedControlPlane"="mettwo" "AWSManagedMachinePool"="mettwo-0" "Cluster"="mettwo" "MachinePool"="mettwo-0" "namespace"="default" "cluster"="mettwo" "nodegroup"="default_mettwo-0"
I0820 12:16:50.572883       1 nodegroup.go:451] controllers/AWSManagedMachinePool "msg"="Found owned EKS nodegroup in AWS" "AWSManagedControlPlane"="mettwo" "AWSManagedMachinePool"="mettwo-0" "Cluster"="mettwo" "MachinePool"="mettwo-0" "namespace"="default" "cluster-name"="mettwo" "nodegroup-name"="default_mettwo-0"
I0820 12:16:50.665236       1 nodegroup.go:340] controllers/AWSManagedMachinePool "msg"="Creating taints update for node group" "AWSManagedControlPlane"="mettwo" "AWSManagedMachinePool"="mettwo-0" "Cluster"="mettwo" "MachinePool"="mettwo-0" "namespace"="default" "name"="default_mettwo-0" "num_current"=0 "num_required"=0
I0820 12:16:50.665284       1 nodegroup.go:371] controllers/AWSManagedMachinePool "msg"="No updates required for node group taints" "AWSManagedControlPlane"="mettwo" "AWSManagedMachinePool"="mettwo-0" "Cluster"="mettwo" "MachinePool"="mettwo-0" "namespace"="default" "name"="default_mettwo-0"
E0820 12:16:50.665870       1 runtime.go:78] Observed a panic: "invalid memory address or nil pointer dereference" (runtime error: invalid memory address or nil pointer dereference)
goroutine 450 [running]:
k8s.io/apimachinery/pkg/util/runtime.logPanic(0x216f600, 0x3e8e210)
	/go/pkg/mod/k8s.io/apimachinery@v0.17.17/pkg/util/runtime/runtime.go:74 +0xa3
k8s.io/apimachinery/pkg/util/runtime.HandleCrash(0x0, 0x0, 0x0)
	/go/pkg/mod/k8s.io/apimachinery@v0.17.17/pkg/util/runtime/runtime.go:48 +0x82
panic(0x216f600, 0x3e8e210)
	/usr/local/go/src/runtime/panic.go:679 +0x1b2
sigs.k8s.io/cluster-api-provider-aws/pkg/cloud/services/eks.(*NodegroupService).reconcileNodegroupConfig(0xc000811aa0, 0xc0004ae0f0, 0x0, 0x0)
	/workspace/pkg/cloud/services/eks/nodegroup.go:409 +0x400
sigs.k8s.io/cluster-api-provider-aws/pkg/cloud/services/eks.(*NodegroupService).reconcileNodegroup(0xc000811aa0, 0xc0008f9440, 0xc0006a36e8)
	/workspace/pkg/cloud/services/eks/nodegroup.go:473 +0x3d1
sigs.k8s.io/cluster-api-provider-aws/pkg/cloud/services/eks.(*NodegroupService).ReconcilePool(0xc000811aa0, 0xc000811aa0, 0xc000056070)
	/workspace/pkg/cloud/services/eks/eks.go:100 +0x186
sigs.k8s.io/cluster-api-provider-aws/exp/controllers.(*AWSManagedMachinePoolReconciler).reconcileNormal(0xc000758230, 0x2a4e3a0, 0xc000056068, 0xc0007cc380, 0xc000400680, 0xc00045ac00, 0xc0008f9440, 0xc0001ed200)
	/workspace/exp/controllers/awsmanagedmachinepool_controller.go:190 +0x1a2
sigs.k8s.io/cluster-api-provider-aws/exp/controllers.(*AWSManagedMachinePoolReconciler).Reconcile(0xc000758230, 0xc0007e67c8, 0x7, 0xc0007e67c0, 0x8, 0xc000a14e00, 0x0, 0x0, 0x0)
	/workspace/exp/controllers/awsmanagedmachinepool_controller.go:174 +0x933
sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).reconcileHandler(0xc0000b4420, 0x223efe0, 0xc0009edcc0, 0xc0004de500)
	/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.5.14/pkg/internal/controller/controller.go:255 +0x162
sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).processNextWorkItem(0xc0000b4420, 0x0)
	/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.5.14/pkg/internal/controller/controller.go:231 +0xcb
sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).worker(0xc0000b4420)
	/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.5.14/pkg/internal/controller/controller.go:210 +0x2b
k8s.io/apimachinery/pkg/util/wait.JitterUntil.func1(0xc000730230)
	/go/pkg/mod/k8s.io/apimachinery@v0.17.17/pkg/util/wait/wait.go:152 +0x5e
k8s.io/apimachinery/pkg/util/wait.JitterUntil(0xc000730230, 0x3b9aca00, 0x0, 0x1, 0xc000088240)
	/go/pkg/mod/k8s.io/apimachinery@v0.17.17/pkg/util/wait/wait.go:153 +0xf8
k8s.io/apimachinery/pkg/util/wait.Until(0xc000730230, 0x3b9aca00, 0xc000088240)
	/go/pkg/mod/k8s.io/apimachinery@v0.17.17/pkg/util/wait/wait.go:88 +0x4d
created by sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).Start.func1
	/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.5.14/pkg/internal/controller/controller.go:192 +0x468
panic: runtime error: invalid memory address or nil pointer dereference [recovered]
	panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x8 pc=0x1ddcbf0]

goroutine 450 [running]:
k8s.io/apimachinery/pkg/util/runtime.HandleCrash(0x0, 0x0, 0x0)
	/go/pkg/mod/k8s.io/apimachinery@v0.17.17/pkg/util/runtime/runtime.go:55 +0x105
panic(0x216f600, 0x3e8e210)
	/usr/local/go/src/runtime/panic.go:679 +0x1b2
sigs.k8s.io/cluster-api-provider-aws/pkg/cloud/services/eks.(*NodegroupService).reconcileNodegroupConfig(0xc000811aa0, 0xc0004ae0f0, 0x0, 0x0)
	/workspace/pkg/cloud/services/eks/nodegroup.go:409 +0x400
sigs.k8s.io/cluster-api-provider-aws/pkg/cloud/services/eks.(*NodegroupService).reconcileNodegroup(0xc000811aa0, 0xc0008f9440, 0xc0006a36e8)
	/workspace/pkg/cloud/services/eks/nodegroup.go:473 +0x3d1
sigs.k8s.io/cluster-api-provider-aws/pkg/cloud/services/eks.(*NodegroupService).ReconcilePool(0xc000811aa0, 0xc000811aa0, 0xc000056070)
	/workspace/pkg/cloud/services/eks/eks.go:100 +0x186
sigs.k8s.io/cluster-api-provider-aws/exp/controllers.(*AWSManagedMachinePoolReconciler).reconcileNormal(0xc000758230, 0x2a4e3a0, 0xc000056068, 0xc0007cc380, 0xc000400680, 0xc00045ac00, 0xc0008f9440, 0xc0001ed200)
	/workspace/exp/controllers/awsmanagedmachinepool_controller.go:190 +0x1a2
sigs.k8s.io/cluster-api-provider-aws/exp/controllers.(*AWSManagedMachinePoolReconciler).Reconcile(0xc000758230, 0xc0007e67c8, 0x7, 0xc0007e67c0, 0x8, 0xc000a14e00, 0x0, 0x0, 0x0)
	/workspace/exp/controllers/awsmanagedmachinepool_controller.go:174 +0x933
sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).reconcileHandler(0xc0000b4420, 0x223efe0, 0xc0009edcc0, 0xc0004de500)
	/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.5.14/pkg/internal/controller/controller.go:255 +0x162
sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).processNextWorkItem(0xc0000b4420, 0x0)
	/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.5.14/pkg/internal/controller/controller.go:231 +0xcb
sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).worker(0xc0000b4420)
	/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.5.14/pkg/internal/controller/controller.go:210 +0x2b
k8s.io/apimachinery/pkg/util/wait.JitterUntil.func1(0xc000730230)
	/go/pkg/mod/k8s.io/apimachinery@v0.17.17/pkg/util/wait/wait.go:152 +0x5e
k8s.io/apimachinery/pkg/util/wait.JitterUntil(0xc000730230, 0x3b9aca00, 0x0, 0x1, 0xc000088240)
	/go/pkg/mod/k8s.io/apimachinery@v0.17.17/pkg/util/wait/wait.go:153 +0xf8
k8s.io/apimachinery/pkg/util/wait.Until(0xc000730230, 0x3b9aca00, 0xc000088240)
	/go/pkg/mod/k8s.io/apimachinery@v0.17.17/pkg/util/wait/wait.go:88 +0x4d
created by sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).Start.func1
	/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.5.14/pkg/internal/controller/controller.go:192 +0x468

```

**What did you expect to happen:**
No SIGSEGV

**Anything else you would like to add:**
The issue appeared after upgrade from v0.6.5 to v0.6.8, so I tried to build a new management cluster with latest v1alpha3 components with the same result. The feature flag EXP_EKS must be set to true:
```
export EXP_EKS=true
export EXP_EKS_IAM=true
export EXP_EKS_ADD_ROLES=true
export EXP_MACHINE_POOL=true
```

**Environment:**

- Cluster-api-provider-aws version: v0.6.8
- Kubernetes version: (use `kubectl version`): Server Version: version.Info{Major:"1", Minor:"19+", GitVersion:"v1.19.13-eks-8df270", GitCommit:"8df2700a72a2598fa3a67c05126fa158fd839620", GitTreeState:"clean", BuildDate:"2021-07-31T01:36:57Z", GoVersion:"go1.15.14", Compiler:"gc", Platform:"linux/amd64"}
- OS (e.g. from `/etc/os-release`):

**Comment by thof** [comment]
It seems the regression was introduced in v0.6.6.

**Comment by richardcase** [comment]
@thof - could you share your `AWSManagedMachonePool` definition?

**Comment by thof** [comment]
@richardcase 
```
bash-5.1$ kubectl get awsmanagedmachinepool.infrastructure.cluster.x-k8s.io/mettwo-0 -o yaml
apiVersion: infrastructure.cluster.x-k8s.io/v1alpha3
kind: AWSManagedMachinePool
metadata:
  annotations:
    kubectl.kubernetes.io/last-applied-configuration: |
      {"apiVersion":"infrastructure.cluster.x-k8s.io/v1alpha3","kind":"AWSManagedMachinePool","metadata":{"annotations":{},"name":"mettwo-0","namespace":"default"},"spec":{"amiType":"AL2_x86_64","instanceType":"t3.large"}}
  creationTimestamp: "2021-08-19T13:24:44Z"
  finalizers:
  - awsmanagedmachinepools.infrastructure.cluster.x-k8s.io
  generation: 1
  labels:
    cluster.x-k8s.io/cluster-name: mettwo
  name: mettwo-0
  namespace: default
  ownerReferences:
  - apiVersion: exp.cluster.x-k8s.io/v1alpha3
    blockOwnerDeletion: true
    controller: true
    kind: MachinePool
    name: mettwo-0
    uid: 13a221de-3fde-447c-a1be-b5c964057f13
  resourceVersion: "459526"
  selfLink: /apis/infrastructure.cluster.x-k8s.io/v1alpha3/namespaces/default/awsmanagedmachinepools/mettwo-0
  uid: ae26ed4e-a33a-4b64-ac48-3a4bfdddd935
spec:
  amiType: AL2_x86_64
  eksNodegroupName: default_mettwo-0
  instanceType: t3.large
  providerIDList:
  - aws://us-east-2b/i-02f75ba5032d5e7a5
  - aws://us-east-2c/i-05ddd00586a26b221
  - aws://us-east-2b/i-0652ca197719d59c0
  roleName: mettwo-default_mettwo-0-nodegroup-iam-service-role
status:
  conditions:
  - lastTransitionTime: "2021-08-19T13:37:40Z"
    status: "True"
    type: Ready
  - lastTransitionTime: "2021-08-20T12:45:49Z"
    status: "True"
    type: EKSNodegroupReady
  - lastTransitionTime: "2021-08-19T13:37:40Z"
    status: "True"
    type: IAMNodegroupRolesReady
  ready: true
  replicas: 3
```

**Comment by richardcase** [comment]
Thanks. I will try this out.

**Comment by richardcase** [comment]
@thof - did you use a a config file when doing `clusterawsadm bootstrap iam create-cloud-formation-stack`? If you did would you also be able to share that?

**Comment by thof** [comment]
@richardcase yes, I did `clusterawsadm bootstrap iam create-cloudformation-stack --config cluster-api/bootstrap-config.yaml`:
```
apiVersion: bootstrap.aws.infrastructure.cluster.x-k8s.io/v1alpha1
kind: AWSIAMConfiguration
spec:
  eks:
    enable: true
    iamRoleCreation: true # Set to true if you plan to use the EKSEnableIAM feature flag to enable automatic creation of IAM roles
    defaultControlPlaneRole:
      disable: false # Set to false to enable creation of the default control plane role
    managedMachinePool:
      disable: false # Set to false to enable creation of the default node role for managed machine pools
```

**Comment by richardcase** [comment]
Thanks, will use that to investigate.

**Comment by thof** [comment]
@richardcase I edited my previous comment with env vars.

**Comment by richardcase** [comment]
> @richardcase I edited my previous comment with env vars.

Thanks @thof - i'm using them both now to see if i can re-create the issue. I'll let you know how i get on....

**Comment by thof** [comment]
@richardcase any luck with reproducing this bug?

**Comment by richardcase** [comment]
Not yet, i'm afraid. Still working on it

**Comment by thof** [comment]
Ok, let me share more info with you within couple minutes.

**Comment by thof** [comment]
@richardcase here is the guide with `metone-engineering.yaml` content below.

Create temporary bootstrap local cluster:
```
kind create cluster
```

#### Configure AWS provider

To install [AWS provider](https://cluster-api-aws.sigs.k8s.io/) run:
```sh
export AWS_REGION=us-east-2 # This is used to help encode your environment variables
export AWS_ACCESS_KEY_ID=<your-access-key>
export AWS_SECRET_ACCESS_KEY=<your-secret-access-key>

clusterawsadm bootstrap iam create-cloudformation-stack --config cluster-api/bootstrap-config.yaml

export AWS_B64ENCODED_CREDENTIALS=$(clusterawsadm bootstrap credentials encode-as-profile)

# Finally, initialize the management cluster
# https://cluster-api-aws.sigs.k8s.io/topics/eks/enabling.html
export EXP_EKS=true
export EXP_EKS_IAM=true
export EXP_EKS_ADD_ROLES=true
export EXP_MACHINE_POOL=true

clusterctl init --infrastructure=aws --control-plane aws-eks --bootstrap aws-eks
```

Create and deploy ssh key:
```sh
aws ssm put-parameter --name "/sigs.k8s.io/cluster-api-provider-aws/ssh-key" \
  --type SecureString \
  --value "$(aws ec2 create-key-pair --key-name cluster-api | jq .KeyMaterial -r)"

export AWS_SSH_KEY_NAME=cluster-api
```

Create target metacluster `metone`:
```
k apply -f cluster-api/live/metone-engineering.yaml
```

**Once the metone is ready I already see crashing capa-controller-manager in kind cluster**

```
clusterctl get kubeconfig metone > kubeconfig-metone.yaml
clusterctl init --infrastructure=aws --control-plane aws-eks --bootstrap aws-eks --kubeconfig=kubeconfig-metone.yaml
clusterctl move --to-kubeconfig=kubeconfig-metone.yaml
```



The content of `metone-engineering.yaml` https://gist.github.com/thof/dd074c81bdbb366b678b2756ac8b0877

**Comment by richardcase** [comment]
Fantastic, thanks @thof

**Comment by richardcase** [comment]
I haven't been able to reproduce but i think i see the issue......

**Comment by thof** [comment]
Interesting... I'm switching between versions and when I downgrade to v0.6.5 the pod runs fine but when I upgrade to v0.6.6 the CrashloopBackOff is immediately present.

**Comment by richardcase** [comment]
There was a change (#2407) that was added in 0.6.6 in this area of code....actually by me :roll_eyes: 

I will have a PR for a fix shortly.

**Comment by thof** [comment]
I tried to build the container image from your fork with backport https://github.com/richardcase/cluster-api-provider-aws/tree/scaling_config_06 using `REGISTRY=docker.io/my-reg make docker-build` and all went fine. However when I used the image in `capa-controller-manager` deployment it started to fail on startup complaining that `flag provided but not defined: -feature-gates`. I'm not sure what went wrong but it seems the local build guide is missing some important information.

**Comment by richardchen-db** [comment]
@richardcase Thank you for fixing it! It looks like the fix is not backported to [release 0.7](https://github.com/kubernetes-sigs/cluster-api-provider-aws/blob/release-0.7/pkg/cloud/services/eks/nodegroup.go#L411). Should I create a separate issue to track this?

**Comment by richardcase** [comment]
> @richardcase Thank you for fixing it! It looks like the fix is not backported to [release 0.7](https://github.com/kubernetes-sigs/cluster-api-provider-aws/blob/release-0.7/pkg/cloud/services/eks/nodegroup.go#L411). Should I create a separate issue to track this?

As discussed on slack this will be available in 1.0.0.
